subsys/net/buf.c

```diff
@@ -91,23 +79,161 @@ void net_buf_reset(struct net_buf *buf)
 	NET_BUF_ASSERT(buf->flags == 0);
 	NET_BUF_ASSERT(buf->frags == NULL);
 
-	buf->len   = 0;
-	buf->data  = buf->__buf;
+	net_buf_simple_reset(&buf->b);
+}
+
+static u8_t *generic_data_ref(struct net_buf *buf, u8_t *data)
+{
+	u8_t *ref_count;
+
+	ref_count = data - 1;
+	(*ref_count)++;
+
+	return data;
+}
+
+static u8_t *mem_pool_data_alloc(struct net_buf *buf, size_t size,
+				 s32_t timeout)
+{
+	struct net_buf_pool *buf_pool = net_buf_pool_get(buf->pool_id);
+	struct k_mem_pool *pool = buf_pool->alloc->alloc_data;
+	struct k_mem_block block;
+	u8_t *ref_count;
+
+	size += sizeof(struct k_mem_block_id) + 1;
+	if (k_mem_pool_alloc(pool, &block, size, timeout)) {
+		return NULL;
+	}
+
+	/* save the block descriptor info at the start of the actual block */
+	memcpy(block.data, &block.id, sizeof(block.id));
+
+	ref_count = (u8_t *)block.data + sizeof(block.id);
+	*ref_count = 1;
+
+	/* Return pointer to the byte following the ref count */
+	return ref_count + 1;
+}
+
+static void mem_pool_data_unref(struct net_buf *buf, u8_t *data)
+{
+	struct k_mem_block_id id;
+	u8_t *ref_count;
+
+	ref_count = data - 1;
+	if (--(*ref_count)) {
+		return;
+	}
+
+	/* Need to copy to local variable due to alignment */
+	memcpy(&id, ref_count - sizeof(id), sizeof(id));
+	k_mem_pool_free_id(&id);
+}
+
+const struct net_buf_data_cb net_buf_var_cb = {
+	.alloc = mem_pool_data_alloc,
+	.ref   = generic_data_ref,
+	.unref = mem_pool_data_unref,
+};
+
+static u8_t *fixed_data_alloc(struct net_buf *buf, size_t size, s32_t timeout)
+{
+	struct net_buf_pool *pool = net_buf_pool_get(buf->pool_id);
+	const struct net_buf_pool_fixed *fixed = pool->alloc->alloc_data;
+
+	if (size > fixed->data_size) {
+		return NULL;
+	}
+
+	return fixed->data_pool + fixed->data_size * net_buf_id(buf);
+}
+
+static void fixed_data_unref(struct net_buf *buf, u8_t *data)
+{
+	/* Nothing needed for fixed-size data pools */
```

Comment:
This sounds a bit strange. If buffer's *data* (vs buffer itself) can be ref counted at all, then it should be ref-counted regardless of the underlying implementation.

Reply:
The data allocation API gives the option for implementations that do not support refcounting. For now I opted for the simplest possible implementation for the fixed pool, where reference counting is not supported. This means that currently when you do a net_buf_clone() on such a buffer a new data portion is allocated and the data is copied. Adding ref-counting support is not too hard: we'd need a bitmap to track which data blocks are allocated and look up free ones using e.g.  __builtin_ffs().

Reply:
> The data allocation API gives the option for implementations that do not support refcounting.

So, one problem with that is that there're just too many options (that's a general problem with this PR as a whole, but let's leave that for "summary" review). For example, I'm not entirely sure is that choice would leak into user-level API, or would be terminated at functions like net_buf_clone().

> For now I opted for the simplest possible implementation for the fixed pool

I don't agree with such a choice either. The fixed-size fragments is what we had before, and should be the baseline of this functionality.Then, it should be converted to offer the same functionality as it had before. And I believe that this refactor noticeably changes its behavior, which I'd recommend against.

Reply:
I'm not completely following. Are you against the fixed-size fragments with or without refcounting? Previously there was no refcounting, so right now this PR maintains the old behavior.

As for too many options, previously you objected to there only being a single option (k_mem_pool), and now that there can be custom allocators there are too many? I guess we have to find some balance mid-way :)

Reply:
> Are you against the fixed-size fragments with or without refcounting? Previously there was no refcounting, so right now this PR maintains the old behavior.

Let me put it this way: does net_buf_clone() *and* net_pkt_clone() with new fixed-size impl behave the same way as they did with the old? If so, ok, but this fact isn't immediately obvious (and as I mentioned, to amount of changes, it's a bit hard to focus on review one topic completely (would take a long time to review entire PR)).

> previously you objected

Well, it's not that I "object", but I try to provide multi-faceted review, pointing out various concerns. As I wrote in the summary, I don't "object", I'm all for, but think you need to give an idea how these changes would be tested. Would be build each and every sample and test with all of 3 options in sanitycheck? Something else?

Reply:
I can't comment for net_pkt since I'm not familiar with that code, but for net_buf the semantics of buffer cloning remain at the moment the same with NET_BUF_POOL_FIXED.

Reply:
I'm not changing anything related to the allocation of the net_buf structs themselves. I.e. they're still managed through a k_lifo like before. It's only the data allocation part that has extra structs etc.